### PR TITLE
Hash class properties in CodeHasher

### PR DESCRIPTION
**Pull request by namanjain**

### CODE REVIEW

File: tests/test_code_hasher.py [hunk 1]
```diff
@@ -192,6 +192,30 @@ class ClassWithClassMethod2:
         def v(cls):
             return 20
 
+    class ClassWithProperty:
```
Comment (jqmp) [inline tests/test_code_hasher.py:195]: Should these two classes get added to `values`?

File: bionic/code_hasher.py [hunk 2]
```diff
@@ -219,6 +219,19 @@ def _update_hash(self, hash_accumulator, obj, code_context):
...
+        elif isinstance(obj, (property, types.DynamicClassAttribute)):
+            if isinstance(obj, list):
+                type_prefix = TypePrefix.PROPERTY
+            else:
+                type_prefix = TypePrefix.DYNAMIC_CLASS_ATTR
+            # A property is identified using fget, fset, fdel, and doc.
+            members = [obj.fget, obj.fset, obj.fdel, obj.__doc__]
+            add_to_hash(
+                hash_accumulator,
+                type_prefix=TypePrefix.HASH,
```
Comment (jqmp) [inline bionic/code_hasher.py:231]: Should this be `type_prefix=type_prefix`?

File: bionic/code_hasher.py [hunk 3]
```diff
@@ -219,6 +219,19 @@ def _update_hash(self, hash_accumulator, obj, code_context):
                 obj_bytes=self._check_and_hash(obj.value, code_context),
             )
 
+        elif isinstance(obj, (property, types.DynamicClassAttribute)):
+            if isinstance(obj, list):
+                type_prefix = TypePrefix.PROPERTY
+            else:
+                type_prefix = TypePrefix.DYNAMIC_CLASS_ATTR
+            # A property is identified using fget, fset, fdel, and doc.
+            members = [obj.fget, obj.fset, obj.fdel, obj.__doc__]
```
Comment (jqmp) [inline bionic/code_hasher.py:228]: Hmm, do we hash docstrings in other contexts?

File: bionic/code_hasher.py [hunk 4]
```diff
@@ -219,6 +219,19 @@ def _update_hash(self, hash_accumulator, obj, code_context):
                 obj_bytes=self._check_and_hash(obj.value, code_context),
             )
 
+        elif isinstance(obj, (property, types.DynamicClassAttribute)):
+            if isinstance(obj, list):
```
Comment (jqmp) [inline bionic/code_hasher.py:223]: What's the relationship between being a property and being a list?

File: bionic/code_hasher.py [hunk 5]
```diff
@@ -219,6 +219,19 @@ def _update_hash(self, hash_accumulator, obj, code_context):
                 obj_bytes=self._check_and_hash(obj.value, code_context),
             )
 
+        elif isinstance(obj, (property, types.DynamicClassAttribute)):
```
Comment (jqmp) [inline bionic/code_hasher.py:222]: When would we encounter a `types.DynamicClassAttribute`? Is there a test case for this?

File: bionic/code_hasher.py (hunk 2 above)
Comment (namanjain) [reply bionic/code_hasher.py:231]: Yes 🤦🏽

File: bionic/code_hasher.py (hunk 3 above)
Comment (namanjain) [reply bionic/code_hasher.py:228]: We do. It's part of `co_consts` for routines and we get it as a `member` for class types.

File: bionic/code_hasher.py (hunk 4 above)
Comment (namanjain) [reply bionic/code_hasher.py:223]: This should have been a property 🤦🏽 . I'll see if I can add a test where property is replaced with dynamic class attr.

File: bionic/code_hasher.py (hunk 5 above)
Comment (namanjain) [reply bionic/code_hasher.py:222]: `Enum`s use it and we hash `TypePrefix` that tests for it. But related to [this comment](https://github.com/square/bionic/pull/329#discussion_r571264494), I'll see how hard it is to create a `DynamicClassAttribute` and test it directly (like property tests).

File: tests/test_code_hasher.py (hunk 1 above)
Comment (namanjain) [reply tests/test_code_hasher.py:195]: Yes 🤦🏽

File: bionic/code_hasher.py (hunk 2 above)
Comment (namanjain) [reply bionic/code_hasher.py:231]: ah with the current pattern, this should be `TypePrefix.Hash`. But I'm missing `add_to_hash(hash_accumulator, type_prefix=type_prefix)` before this.

File: bionic/code_hasher.py (hunk 4 above)
Comment (namanjain) [reply bionic/code_hasher.py]: I added a test for it and it fails without fixing this bug. Good catch!

**Review by jqmp — APPROVED** [review]

**Pull request merged**